### Beginning of the Assignment - exploration

In [ ]:
import pandas as pd
import numpy as np
import seaborn as sns
#pd.set_option('max_columns', 200)

### PHIL DATA SET

In [2]:
character_nicknames_df = pd.read_csv('datasets/PHIL DATA SET/character_nicknames.csv')

In [3]:
details_df = pd.read_csv('datasets/PHIL DATA SET/details.csv')

In [4]:
favs_df = pd.read_csv('datasets/PHIL DATA SET/favs.csv')


In [5]:
person_details_df = pd.read_csv('datasets/PHIL DATA SET/person_details.csv')


In [6]:
person_alternate_names_df = pd.read_csv('datasets/PHIL DATA SET/person_alternate_names.csv')


In [7]:
person_anime_works_df = pd.read_csv('datasets/PHIL DATA SET/person_anime_works.csv')


In [8]:
stats_df = pd.read_csv('datasets/PHIL DATA SET/stats.csv')


### DENIS DATA SET

In [9]:
ratings_df = pd.read_csv('datasets/DENIS DATA SET/ratings.csv')

In [10]:
characters_df = pd.read_csv('datasets/DENIS DATA SET/characters.csv')

In [11]:
character_anime_works_df = pd.read_csv('datasets/DENIS DATA SET/character_anime_works.csv')

In [12]:
person_voice_works_df = pd.read_csv('datasets/DENIS DATA SET/person_voice_works.csv')

In [13]:
profiles_df = pd.read_csv('datasets/DENIS DATA SET/profiles.csv')

In [14]:
recommendations_df = pd.read_csv('datasets/DENIS DATA SET/recommendations.csv')

# GUIDELINES
### BEFORE STARTING
Use Conda in order to do the correct setup

When we deliver the project, we need to tell the lecturer to run the jupyter notebook before running the TWEB part. How? Write it in the Report
Give instructions in order to make it run properly

#TODO to write on the report:
Which are the cool things, problems, map names not allined so we had to normalize them and so on

Things to do for each dataset:
1. Give it a look with .head and/or .tail
2. .describe and check if all the numeric values make sense (e.g. year=300 makes no sense in our context)
3. Check the format: objects to date if we need it. Check if all the dates are in the same format: us or eu
    also check if there's any 29/02/2013. It doesn't exist right? Maybe this is too much lol
4. Check for duplicates
5. CHECK FOR CORRELATION: df.corr() (e.g. with longer duration, there are more actors)
#TODO
6. ADD or Remove columns?
7. GROUPBY selects the elements and makes group out of it, combines the numeric fields of each specific group
#TODO We could use it grouping for language and looking at how many anime are made in japan, stating it's the first country where the culture of doing (and watching) anime is SO big
8. Aggregations: we can apply multiple different aggregated functions (e.g. for the first column you sum the data, for the second you do the average and so on)
9. Transformations: apply operations and return results aligned with the original DF

10. Removing NaNs is wrong in general because Pandas will skip it.
We do it when? Is it safe to remove NaNs rows if EVERY field in the row is empty? I hope so lol
BE CAREFUL if they are foreign keys: for example, if a person has a nan in the "anime he worked in" field, it shouldn't be dropped
NEVER replace with invalid values (e.g. -1)
IF we use df.dropna(subset=["name"],inplace=True)
the inplace means that the df itself is modified and will result in the one without the na. Without "inplace=true" you'll need to assign the result to another df (or the same) 


Proviamo i plot? df.plot()

11. Check if data are consistent (e.g. normalizing names of countries and/or numeric fields, describing them and checking what they are)

12. Normalize data types all in the same place (e.g. all the dates in the same cell)


BONUS: NEVER USE LOOP FOR, NEVER DUPLICATE DATA (unless necessary)


##### First look

In [15]:
character_nicknames_df.shape

(37080, 2)

In [16]:
character_nicknames_df.head()

character_mal_id              nickname
0            280205              Hikaruko
1            280129              Hinacchi
2            280127         Bertha Willis
3            280066                 Jimmy
4            280059  Full Body Red Square

In [17]:
character_nicknames_df.columns
#will list all the columns. Not necessary here but kept as a pattern to follow with the following files

Index(['character_mal_id', 'nickname'], dtype='object')

In [18]:
character_nicknames_df.dtypes

character_mal_id     int64
nickname            object
dtype: object

### Data preparation (cleaning)


##### On the first dataset we may need to check for duplicates.
What does that mean? We have 102 rows that are duplicated over a 37080 rows dataset.
Why is that? Are there some characters that have multiple nicknames, so they are repeated in the dataset?

In [ ]:
character_nicknames_df.info()

In [ ]:
character_nicknames_df.describe(include='all')   

In [19]:
character_nicknames_df.loc[character_nicknames_df.duplicated()]
#by default will give us the second

character_mal_id        nickname
328              276501            Koko
329              276501         Cao Cao
352              275883            King
356              275835  Eldest Brother
447              274437   The Half-Fool
...                 ...             ...
37051            281206          Apemon
37052            281206          Apemon
37053            281206          Apemon
37054            281206          Apemon
37068            281717         Colosso

[102 rows x 2 columns]

Mhh they're somehow different so yeah, the same character could have different nicknames.
We want to drop though the ones that are exactly the same.

In [20]:
#this way we drop the duplicates on the first dataset

# OLD VERSION 
# character_nicknames_df = character_nicknames_df.loc[~character_nicknames_df.duplicated()].reset_index(drop=True).copy()

#we don't need to use a subset here because there are just 2 columns

#why don't we just use drop_duplicates()?
character_nicknames_df.drop_duplicates(keep='first', inplace=True)

Let's check for nan values

In [21]:
character_nicknames_df[character_nicknames_df.isna().any(axis=1)]


character_mal_id nickname
696              271183      NaN
3170             248157      NaN
4880             230372      NaN
15899            135548      NaN
18598            107631      NaN
24278             48077      NaN
25041             42325      NaN
26604             35444      NaN
30115             19532      NaN
30116             19531      NaN
30118             19527      NaN
30214             19265      NaN
30294             18984      NaN
30304             18952      NaN
31320             15681      NaN
33951              6562      NaN

In [22]:
#cleaning the df from nan values
character_nicknames_df.dropna(inplace=True)


## SECOND DATASET

##### On the second dataset we may need to check for missing values and/or inconsistent values, since there are no duplicates

In [23]:
details_df
#anime details
#japanes title could be dropped?
#members stand for how many users have this anime added to their list.
#explicit_genres is empty so can be removed
#licensor and streaming are mostly empty. Do we care?

mal_id                                              title  \
0       59356                                           -Socket-   
1       56036                                             ......   
2        2928                               .hack//G.U. Returner   
3        3269                                .hack//G.U. Trilogy   
4        4469                   .hack//G.U. Trilogy: Parody Mode   
...       ...                                                ...   
28950   59421  Zutaboro Reijou wa Ane no Moto Konyakusha ni D...   
28951   31245  Zutto Mae kara Suki deshita. Kokuhaku Jikkou I...   
28952   36305  Zutto Mae kara Suki deshita. Kokuhaku Jikkou I...   
28953   34895                                   Zutto Suki Datta   
28954   56606                                    Zutto Tomodachi   

                         title_japanese  \
0                              -socket-   
1                                ......   
2                  .HACK//G.U. RETURNER   
3                   .hack//G.U. Trilogy   
4                   .hack//G.U. Trilogy   
...                                 ...   
28950               ずたぼろ令嬢は姉の元婚約者に溺愛される   
28951             ずっと前から好きでした。～告白実行委員会～   
28952  ずっと前から好きでした。～告白実行委員会～ 「金曜日のおはよう」   
28953                          ずっと好きだった   
28954                             ずっと友達   

                                                     url  \
0           https://myanimelist.net/anime/59356/-Socket-   
1                  https://myanimelist.net/anime/56036/-   
2      https://myanimelist.net/anime/2928/hack__GU_Re...   
3      https://myanimelist.net/anime/3269/hack__GU_Tr...   
4      https://myanimelist.net/anime/4469/hack__GU_Tr...   
...                                                  ...   
28950  https://myanimelist.net/anime/59421/Zutaboro_R...   
28951  https://myanimelist.net/anime/31245/Zutto_Mae_...   
28952  https://myanimelist.net/anime/36305/Zutto_Mae_...   
28953  https://myanimelist.net/anime/34895/Zutto_Suki...   
28954  https://myanimelist.net/anime/56606/Zutto_Tomo...   

                                               image_url     type  \
0      https://cdn.myanimelist.net/images/anime/1043/...    Movie   
1      https://cdn.myanimelist.net/images/anime/1057/...    Music   
2      https://cdn.myanimelist.net/images/anime/1798/...      OVA   
3      https://cdn.myanimelist.net/images/anime/1566/...    Movie   
4      https://cdn.myanimelist.net/images/anime/10/86...  Special   
...                                                  ...      ...   
28950  https://cdn.myanimelist.net/images/anime/1518/...       TV   
28951  https://cdn.myanimelist.net/images/anime/3/821...    Movie   
28952  https://cdn.myanimelist.net/images/anime/6/883...  Special   
28953  https://cdn.myanimelist.net/images/anime/1498/...      OVA   
28954  https://cdn.myanimelist.net/images/anime/1462/...    Music   

                status  score  scored_by                 start_date  ...  \
0      Finished Airing    NaN        NaN  2010-01-01T00:00:00+00:00  ...   
1      Finished Airing   6.53      503.0  2023-06-11T00:00:00+00:00  ...   
2      Finished Airing   6.65     9745.0  2007-01-18T00:00:00+00:00  ...   
3      Finished Airing   7.06    15373.0  2007-12-22T00:00:00+00:00  ...   
4      Finished Airing   6.35     4317.0  2008-03-25T00:00:00+00:00  ...   
...                ...    ...        ...                        ...  ...   
28950  Finished Airing   7.37    15624.0  2025-07-05T00:00:00+00:00  ...   
28951  Finished Airing   7.20   104106.0  2016-04-23T00:00:00+00:00  ...   
28952  Finished Airing   7.17    10038.0  2016-10-26T00:00:00+00:00  ...   
28953  Finished Airing   5.68     1887.0  2017-04-21T00:00:00+00:00  ...   
28954  Finished Airing    NaN        NaN  1996-06-01T00:00:00+00:00  ...   

      demographics       source                     rating  episodes  season  \
0               []     Original               G - All Ages       1.0     NaN   
1               []     Original  PG-13 - Teens 13 

In [24]:
# we want to see what are "type"
details_df['type'].value_counts()

type
TV            8414
Movie         4915
OVA           4184
ONA           4096
Music         3999
Special       1755
TV Special     767
CM             483
PV             275
Name: count, dtype: int64

In [25]:
details_df.loc[details_df.duplicated(subset=['url'], keep='first')]

Empty DataFrame
Columns: [mal_id, title, title_japanese, url, image_url, type, status, score, scored_by, start_date, end_date, synopsis, rank, popularity, members, favorites, genres, studios, themes, demographics, source, rating, episodes, season, year, producers, explicit_genres, licensors, streaming]
Index: []

[0 rows x 29 columns]

In [26]:
details_df.query('year>2025')

mal_id                                              title  \
134     61637      29-sai Dokushin Chuuken Boukensha no Nichijou   
733     62000  Akuyaku Reijou wa Ringoku no Outaishi ni Dekia...   
1152    61333                 Ao no Miburo: Serizawa Ansatsu-hen   
1355    60255                                    Arne no Jikenbo   
3361    60509                                 Champignon no Majo   
4687    59853                         Dark Moon: Tsuki no Saidan   
4700    58886                                       Darwin Jihen   
4762    61196                                       Dead Account   
5459    61325                Douse, Koishite Shimaunda. Season 2   
5929    59229             Enen no Shouboutai: San no Shou Part 2   
5985    60151                                     Eris no Seihai   
6212    55830                                  Fate/strange Fake   
7580    55772                          Golden Kamuy: Saishuushou   
7658    61886                       Goumon Baito-kun no Nichijou   
8313    58861                          Hanazakari no Kimitachi e   
8671    60460  Hell Mode: Yarikomizuki no Gamer wa Hai Settei...   
8967    61931  Higeki no Genkyou to Naru Saikyou Gedou Last B...   
9062    61084                    Hikuidori: Ushuu Boro Tobi-gumi   
9080    60028                      Himekishi wa Barbaroi no Yome   
9083    58524         Himesama "Goumon" no Jikan desu 2nd Season   
9383    57466  Honzuki no Gekokujou: Shisho ni Naru Tame ni w...   
10020   58788                                        Ikoku Nikki   
10310   60226                  Isekai no Sata wa Shachiku Shidai   
10657   61608                   Jigoku Sensei Nube (2025) Part 2   
10670   55825                              Jigokuraku 2nd Season   
10731   62432           Jingai Kyoushitsu no Ningengirai Kyoushi   
10970   57658                          Jujutsu Kaisen 3rd Season   
11494   61186        Kamiina Botan, Yoeru Sugata wa Yuri no Hana   
12543   61782                        Kirei ni Shitemoraemasu ka.   
13004   60852                                   Koori no Jouheki   
14707   60310                 Mairimashita! Iruma-kun 4th Season   
14941   62048                                                Mao   
15129   58505                             Mato Seihei no Slave 2   
15238   61335                                Medalist 2nd Season   
15564   60602                                MF Ghost 3rd Season   
17375   59393                                   Niwatori Fighter   
17675   60071                Odayaka Kizoku no Kyuuka no Susume.   
18430   56876  Otonari no Tenshi-sama ni Itsunomanika Dame Ni...   
20144   60153                                          Rilakkuma   
20594   61831  Saikyou no Ousama, Nidome no Jinsei wa Nani wo...   
21062   60371                          Seihantai na Kimi to Boku   
21757   59711                      Shibou Yuugi de Meshi wo Kuu.   
23117   59978                       Sousou no Frieren 2nd Season   
23969   59047                             Tamon-kun Ima Docchi!?   
24333   53913                       Tensei shitara Ken deshita 2   
25217   60395  Toumei Otoko to Ningen Onna: Sonouchi Fuufu ni...   
25355   59983               Tsue to Tsurugi no Wistoria Season 2   
25968   61663                           Uruwashi no Yoi no Tsuki   
26159   61942  Vigilante: Boku no Hero Academia Illegals 2nd ...   
27998   61830                    Yoroi Shin Den Samurai Troopers   
28499   61217                                     Yuusha no Kuzu   
28502   61128             Yuusha Party wo Oidasareta Kiyoubinbou   
28514   56009  Yuusha-kei ni Shosu: Choubatsu Yuusha 9004-tai...   

                                  title_japanese  \
134                                29歳独身中堅冒険者の日常   
733                            悪役令嬢は隣国の王太子に溺愛される   
1152                                 青のミブロ 芹沢暗殺編   
1355                                     アルネの事件簿   
3361                                  シャンピニオンの魔女   
4687                  

In [27]:
details_df.describe()

mal_id         score     scored_by          rank    popularity  \
count  28955.000000  18882.000000  1.888200e+04  21997.000000  28955.000000   
mean   33977.521948      6.390500  2.996308e+04  11033.157794  14500.358798   
std    19616.858566      0.892045  1.219666e+05   6388.279473   8373.663464   
min        1.000000      1.890000  1.010000e+02      1.000000      1.000000   
25%    15454.000000      5.770000  3.322500e+02   5505.000000   7247.500000   
50%    37969.000000      6.360000  1.528000e+03  11008.000000  14497.000000   
75%    50434.500000      7.030000  1.014575e+04  16515.000000  21754.500000   
max    62590.000000      9.290000  2.979733e+06  22020.000000  28999.000000   

            members      favorites      episodes         year  
count  2.895500e+04   28955.000000  28275.000000  6266.000000  
mean   3.875337e+04     430.848973     14.130080  2010.034153  
std    1.673763e+05    4520.610176     47.161445    13.200708  
min    2.300000e+01       0.000000      1.000000  1961.000000  
25%    2.330000e+02       0.000000      1.000000  2005.000000  
50%    1.077000e+03       1.000000      2.000000  2014.000000  
75%    9.193500e+03      17.000000     13.000000  2020.000000  
max    4.230312e+06  243358.000000   3000.000000  2026.000000

In [28]:
details_df[['start_date','season']].query("season.notna()")
#season can be removed? Do we care about the season? We can "calculate" it from the "start_date" field

start_date  season
10     2006-04-06T00:00:00+00:00  spring
11     2002-04-04T00:00:00+00:00  spring
12     2003-01-09T00:00:00+00:00  winter
19     2024-07-07T00:00:00+00:00  summer
20     2025-01-05T00:00:00+00:00  winter
...                          ...     ...
28906  2001-04-07T00:00:00+00:00  spring
28913  2011-05-18T00:00:00+00:00  spring
28932  1980-04-15T00:00:00+00:00  spring
28939  2012-04-03T00:00:00+00:00  spring
28950  2025-07-05T00:00:00+00:00  summer

[6266 rows x 2 columns]

In [29]:
details_df.query("episodes > 2500")

mal_id    title title_japanese  \
13697    9947  Lan Mao      蓝猫淘气3000问   

                                              url  \
13697  https://myanimelist.net/anime/9947/Lan_Mao   

                                               image_url type  \
13697  https://cdn.myanimelist.net/images/anime/5/275...   TV   

                status  score  scored_by                 start_date  ...  \
13697  Finished Airing   6.06      228.0  1999-10-08T00:00:00+00:00  ...   

      demographics    source         rating  episodes  season  year  \
13697           []  Original  PG - Children    3000.0     NaN   NaN   

                                        producers explicit_genres licensors  \
13697  ['Beijing Sunchime Happy Culture Company']              []        []   

      streaming  
13697        []  

[1 rows x 29 columns]

In [30]:
details_df[["start_date", "end_date"]] = details_df[["start_date", "end_date"]].apply(
    pd.to_datetime, errors="coerce"
)

In [31]:
details_df[["scored_by", "rank", "episodes", "year"]] = (
    details_df[["scored_by", "rank", "episodes", "year"]].astype("Int64")
)


In [32]:
details_df.dtypes
#scored_by, rank, episodes, year can be an int instead of a float
#start and end dates are not objects but dates
#do we need to swap the empty [] with Nan or not? WE should in order to be able to use the .isna() method and other pandas methods


mal_id                           int64
title                           object
title_japanese                  object
url                             object
image_url                       object
type                            object
status                          object
score                          float64
scored_by                        Int64
start_date         datetime64[ns, UTC]
end_date           datetime64[ns, UTC]
synopsis                        object
rank                             Int64
popularity                       int64
members                          int64
favorites                        int64
genres                          object
studios                         object
themes                          object
demographics                    object
source                          object
rating                          object
episodes                         Int64
season                          object
year                             Int64
producers                

### THIRD DATASET

In [33]:
favs_df

username   fav_type      id
0          ishikawas      anime   45649
1          ishikawas      anime   38680
2          ishikawas      anime     795
3          ishikawas      anime   37510
4          ishikawas      anime     820
...              ...        ...     ...
4178742  vincent0607  character     497
4178743  vincent0607  character  118739
4178744  vincent0607  character  188177
4178745  vincent0607  character  141354
4178746  vincent0607  character  195178

[4178747 rows x 3 columns]

In [34]:
# we want to see what are "fav_type"
favs_df['fav_type'].value_counts()

fav_type
character    1598040
anime        1531857
people        862186
company       186664
Name: count, dtype: int64

In [35]:
favs_df.isna().sum()


username    4
fav_type    0
id          0
dtype: int64

In [36]:
favs_df.dtypes

username    object
fav_type    object
id           int64
dtype: object

In [37]:
favs_df.duplicated().sum()
#there are no duplicates


np.int64(0)

### FOURTH DATASET

In [38]:
person_alternate_names_df

person_mal_id             alt_name
0                  1         Seki Mondoya
1                  1                 門戸 開
2                  1         Monto Hiraku
3                  3                 雪野五月
4                 10        Kevin Hatcher
...              ...                  ...
20460          89567              Sydsnap
20461          89567  Queen of Degeneracy
20462          89826                   陳浩
20463          89842              Chidori
20464          89842                 Nobu

[20465 rows x 2 columns]

In [39]:
person_alternate_names_df.dtypes

person_mal_id     int64
alt_name         object
dtype: object

In [40]:
person_alternate_names_df.isna().sum()
person_alternate_names_df[person_alternate_names_df.isna().any(axis=1)]


person_mal_id alt_name
1470            2813      NaN
1662            3380      NaN
2552            7070      NaN
3932           12406      NaN
7829           40897      NaN
8647           44952      NaN
8895           46425      NaN
9899           49829      NaN
11160          53940      NaN
11350          54493      NaN
12922          59875      NaN
13746          62816      NaN
13947          63544      NaN
13960          63593      NaN
14871          67684      NaN
15837          71407      NaN
17787          79001      NaN
18139          80585      NaN

In [41]:
person_alternate_names_df.dropna(inplace=True)



In [42]:
# person_alternate_names_df.loc[person_alternate_names_df['person_mal_id'].duplicated()]
person_alternate_names_df[person_alternate_names_df.duplicated(subset=['person_mal_id','alt_name'], keep=False)].sort_values(['person_mal_id','alt_name'])


person_mal_id                      alt_name
2526            7025                    Fumio Tada
2531            7025                    Fumio Tada
2536            7025                    Fumio Tada
2528            7025               Ichirou Miyoshi
2533            7025               Ichirou Miyoshi
...              ...                           ...
17383          77491  Joshua Ricardo Rocha Jiménez
17386          77514                    Scott Page
17387          77514                    Scott Page
18080          80213           Samantha Carmichael
18082          80213           Samantha Carmichael

[68 rows x 2 columns]

In [43]:
person_alternate_names_df.drop_duplicates(keep='first', inplace=True)

### FIFTH DATASET

In [44]:
person_details_df

person_mal_id                                                url  \
0                  1     https://myanimelist.net/people/1/Tomokazu_Seki   
1                  2   https://myanimelist.net/people/2/Tomokazu_Sugita   
2                  3    https://myanimelist.net/people/3/Satsuki_Yukino   
3                  4        https://myanimelist.net/people/4/Aya_Hirano   
4                  5  https://myanimelist.net/people/5/Kenichi_Suzumura   
...              ...                                                ...   
76694          90011  https://myanimelist.net/people/90011/Nanako_Ki...   
76695          90012         https://myanimelist.net/people/90012/Pamon   
76696          90013  https://myanimelist.net/people/90013/Tomoru_Emoto   
76697          90014        https://myanimelist.net/people/90014/Hirari   
76698          90015  https://myanimelist.net/people/90015/Daisuke_T...   

                   website_url  \
0                          NaN   
1          https://agrs.co.jp/   
2                          NaN   
3         http://ayahirano.jp/   
4      https://intention-k.com   
...                        ...   
76694                      NaN   
76695                      NaN   
76696                      NaN   
76697   https://hirari.2-d.jp/   
76698                      NaN   

                                               image_url               name  \
0      https://cdn.myanimelist.net/images/voiceactors...      Tomokazu Seki   
1      https://cdn.myanimelist.net/images/voiceactors...    Tomokazu Sugita   
2      https://cdn.myanimelist.net/images/voiceactors...     Satsuki Yukino   
3      https://cdn.myanimelist.net/images/voiceactors...         Aya Hirano   
4      https://cdn.myanimelist.net/images/voiceactors...   Kenichi Suzumura   
...                                                  ...                ...   
76694  https://cdn.myanimelist.net/img/sp/icon/apple-...   Nanako Kishimoto   
76695  https://cdn.myanimelist.net/img/sp/icon/apple-...              Pamon   
76696  https://cdn.myanimelist.net/img/sp/icon/apple-...       Tomoru Emoto   
76697  https://cdn.myanimelist.net/img/sp/icon/apple-...             Hirari   
76698  https://cdn.myanimelist.net/img/sp/icon/apple-...  Daisuke Taniguchi   

      given_name family_name                   birthday  favorites  \
0             智一           関  1972-09-08T00:00:00+00:00       6219   
1             智和          杉田  1980-10-11T00:00:00+00:00      47666   
2            さつき         ゆきの  1970-05-25T00:00:00+00:00       1777   
3              綾          平野  1987-10-08T00:00:00+00:00      18374   
4             健一          鈴村  1974-09-12T00:00:00+00:00       5176   
...          ...         ...                        ...        ...   
76694         七子          岸本                        NaN          0   
76695        NaN          파몬                        NaN          0   
76696        ともる          柄本                        NaN          0   
76697        NaN         ひらり                        NaN          0   
76698         大介          谷口                        NaN          0   

             relevant_location  
0              Berlin, Germany  
1             Los Angeles, USA  
2                Madrid, Spain  
3                Paris, France  
4                 Osaka, Japan  
...                        ...  
76694            Mumbai, India  
76695             Tokyo, Japan  
76696             Tokyo, Japan  
76697            Paris, France  
76698  Cape Town, South Africa  

[76699 rows x 10 columns]

In [45]:
person_details_df.loc[person_details_df['person_mal_id'].duplicated()]
#we found that the duplicates differ for the "relevant_location" field, which has no interest for us so we drop the duplicates

person_mal_id                                                url  \
4171            4564  https://myanimelist.net/people/4564/Natsuko_Ta...   
4172            4564  https://myanimelist.net/people/4564/Natsuko_Ta...   
5202            5742  https://myanimelist.net/people/5742/Sarah_Strange   
6018            6638  https://myanimelist.net/people/6638/Toshio_Suzuki   
6019            6638  https://myanimelist.net/people/6638/Toshio_Suzuki   
...              ...                                                ...   
64387          77565  https://myanimelist.net/people/77565/Ena_Nishi...   
64402          77579   https://myanimelist.net/people/77579/Amina_Gaede   
64412          77588  https://myanimelist.net/people/77588/Willie_Ra...   
64414          77589    https://myanimelist.net/people/77589/Brandi_Ray   
64417          77591  https://myanimelist.net/people/77591/Ilka_Pinh...   

                                  website_url  \
4171                                      NaN   
4172                                      NaN   
5202                                      NaN   
6018                                      NaN   
6019                                      NaN   
...                                       ...   
64387                                     NaN   
64402  https://www.instagram.com/amina.gaede/   
64412                                     NaN   
64414                                     NaN   
64417                                     NaN   

                                               image_url               name  \
4171   https://cdn.myanimelist.net/images/voiceactors...  Natsuko Takahashi   
4172   https://cdn.myanimelist.net/images/voiceactors...  Natsuko Takahashi   
5202   https://cdn.myanimelist.net/images/voiceactors...      Sarah Strange   
6018   https://cdn.myanimelist.net/images/voiceactors...      Toshio Suzuki   
6019   https://cdn.myanimelist.net/images/voiceactors...      Toshio Suzuki   
...                                                  ...                ...   
64387  https://cdn.myanimelist.net/img/sp/icon/apple-...      Ena Nishikawa   
64402  https://cdn.myanimelist.net/images/voiceactors...        Amina Gaede   
64412  https://cdn.myanimelist.net/images/voiceactors...     Willie Ray Jr.   
64414  https://cdn.myanimelist.net/images/voiceactors...         Brandi Ray   
64417  https://cdn.myanimelist.net/images/voiceactors...      Ilka Pinheiro   

      given_name family_name                   birthday  favorites  \
4171         ナツコ          高橋                        NaN         19   
4172         ナツコ          高橋                        NaN         19   
5202         NaN         NaN                        NaN         11   
6018          敏夫          鈴木  1948-08-19T00:00:00+00:00        354   
6019          敏夫          鈴木  1948-08-19T00:00:00+00:00        354   
...          ...         ...                        ...        ...   
64387         絵奈          西川                        NaN          0   
64402        NaN         NaN  1998-07-05T00:00:00+00:00          0   
64412        NaN         NaN                        NaN          0   
64414        NaN         NaN                        NaN          0   
64417        NaN         NaN                        NaN          0   

             relevant_location  
4171             Paris, France  
4172   Cape Town, South Africa  
5202           Yokohama, Japan  
6018             Paris, France  
6019             Mumbai, India  
...                        ...  
64387            Paris, France  
64402        São Paulo, Brazil  
64412          Berlin, Germany  
64414            Mumbai, India  
64417            New York, USA  

[105 rows x 10 columns]

In [46]:
person_details_df.drop_duplicates(subset=['person_mal_id', 'url', 'name'], keep='first', inplace=True)


In [47]:
person_details_df.dtypes
#we need to change birthday from object to data

person_mal_id         int64
url                  object
website_url          object
image_url            object
name                 object
given_name           object
family_name          object
birthday             object
favorites             int64
relevant_location    object
dtype: object

In [48]:
person_details_df["birthday"] = pd.to_datetime(person_details_df["birthday"], errors='coerce')

In [49]:
person_details_df["birthday"].min(), person_details_df["birthday"].max()
#makes sense because they're just composers of used music in anime

(Timestamp('1678-03-04 00:00:00+0000', tz='UTC'),
 Timestamp('2023-06-15 00:00:00+0000', tz='UTC'))

In [50]:
person_details_df.isna().sum()
#we have to check the nan values

person_mal_id            0
url                      0
website_url          59186
image_url                0
name                     2
given_name           30211
family_name          18610
birthday             59399
favorites                0
relevant_location        0
dtype: int64

In [51]:
person_details_df[person_details_df["name"].isna()]


person_mal_id                                        url website_url  \
46687          59857  https://myanimelist.net/people/59857/None         NaN   
75920          89106  https://myanimelist.net/people/89106/NULL         NaN   

                                               image_url name given_name  \
46687  https://cdn.myanimelist.net/img/sp/icon/apple-...  NaN        NaN   
75920  https://cdn.myanimelist.net/img/sp/icon/apple-...  NaN        NaN   

      family_name birthday  favorites relevant_location  
46687          のね      NaT          0     Paris, France  
75920         NaN      NaT          0     Paris, France

We can join the two tables person_details_df and person_alternate_names_df having the keys that match.
Putting the alternate names in a new column called alt_name and having a list of those inside

### SIXTH DATASET

In [52]:
person_anime_works_df

person_mal_id                     position  anime_mal_id
0                   1       Theme Song Performance          3080
1                   1    Inserted Song Performance         15699
2                   1  Theme Song Performance (OP)           247
3                   1       Theme Song Performance           258
4                   1  Theme Song Performance (ED)         34825
...               ...                          ...           ...
458086          89951         In-Between Animation         11001
458087          89951                Key Animation         55092
458088          89951     Key Animation (eps 4, 9)         20053
458089          89951                Key Animation         50553
458090          89965          Music Not Yet Aired         62579

[458091 rows x 3 columns]

In [53]:
person_anime_works_df.dtypes
#the types are correct

person_mal_id     int64
position         object
anime_mal_id      int64
dtype: object

In [54]:
person_anime_works_df.isna().sum()
#There's no nan value

person_mal_id    0
position         0
anime_mal_id     0
dtype: int64

### SEVENTH DATASET

In [55]:
stats_df.filter(regex="_votes$").astype("Int64")
stats_df[stats_df.filter(regex="_votes$").columns] = (
    stats_df.filter(regex="_votes$").astype("Int64")
)

In [56]:
stats_df.isna().sum()
#there are 430 series without any votes

mal_id                   0
watching                 0
completed                0
on_hold                  0
dropped                  0
plan_to_watch            0
total                    0
score_1_votes          430
score_1_percentage     430
score_2_votes          430
score_2_percentage     430
score_3_votes          430
score_3_percentage     430
score_4_votes          430
score_4_percentage     430
score_5_votes          430
score_5_percentage     430
score_6_votes          430
score_6_percentage     430
score_7_votes          430
score_7_percentage     430
score_8_votes          430
score_8_percentage     430
score_9_votes          430
score_9_percentage     430
score_10_votes         430
score_10_percentage    430
dtype: int64

### EIGHTH DATASET 

In [57]:
ratings_df

username  anime_id         status  score  is_rewatching  \
0          --------788     30276       watching      7            0.0   
1          --------788     28851      completed      7            0.0   
2          --------788     41168      completed      7            0.0   
3          --------788     22199      completed     10            0.0   
4          --------788     16498      completed     10            0.0   
...                ...       ...            ...    ...            ...   
124298352      arizkim     52305  plan_to_watch      0            0.0   
124298353      arizkim      4224  plan_to_watch      0            0.0   
124298354      arizkim     54790  plan_to_watch      0            0.0   
124298355      arizkim     53835  plan_to_watch      0            0.0   
124298356      arizkim     33352  plan_to_watch      0            0.0   

           num_watched_episodes  
0                             3  
1                             1  
2                             1  
3                            24  
4                            25  
...                         ...  
124298352                     0  
124298353                     0  
124298354                     0  
124298355                     0  
124298356                     0  

[124298357 rows x 6 columns]

In [58]:
ratings_df.columns

Index(['username', 'anime_id', 'status', 'score', 'is_rewatching',
       'num_watched_episodes'],
      dtype='object')

In [59]:
# we have to understand the sense of "num_watched_episodes" and the link with "is_rewatching"
ratings_df.query('is_rewatching == 1')

username  anime_id     status  score  is_rewatching  \
934                   CKK2         6  completed      9            1.0   
8939       FollowYourHeart      1210  completed     10            1.0   
11047         ----Haku----     39017  completed      7            1.0   
22879            KarioBaka     10087  completed      9            1.0   
22928            KarioBaka     14467  completed      0            1.0   
...                    ...       ...        ...    ...            ...   
124291819      arissabelle      2167  completed      0            1.0   
124291878      arissabelle     33352  completed     10            1.0   
124291883      arissabelle      6547    on_hold      0            1.0   
124297299      ariyanroy04     46569  completed      0            1.0   
124297323      ariyanroy04        19  completed      0            1.0   

           num_watched_episodes  
934                           8  
8939                         24  
11047                         3  
22879                         3  
22928                         6  
...                         ...  
124291819                     0  
124291878                     0  
124291883                     1  
124297299                    13  
124297323                    28  

[103341 rows x 6 columns]

In [60]:
ratings_df.dtypes

username                 object
anime_id                  int64
status                   object
score                     int64
is_rewatching           float64
num_watched_episodes      int64
dtype: object

In [61]:
# change "is_rewatching" from float to Int8, to save memory
ratings_df["is_rewatching"] = ratings_df["is_rewatching"].astype("Int8") 

In [62]:
ratings_df[["anime_id","score","num_watched_episodes"]].agg(["min", "max"])


anime_id  score  num_watched_episodes
min         1      0                     0
max     62893     10                 65535

In [63]:
# to save up some memory, we can change "anime_id" from Int64 to Int32 because there are no anime with id > 2,147,483,647
ratings_df["anime_id"] = ratings_df["anime_id"].astype("Int32")

In [64]:
ratings_df[ratings_df.duplicated(subset=['username','anime_id'], keep=False)].sort_values(['username','anime_id'])

username  anime_id    status  score  is_rewatching  \
74586609  Doopelsdoo     59459  watching      0              0   
74616569  Doopelsdoo     59459  watching      0              0   
74594917    Door_mp3     59644  watching      8              0   
74626628    Door_mp3     59644  watching      8              0   
74594935    Door_mp3     60564  watching      8              0   
74626629    Door_mp3     60564  watching      8              0   
74594941    Door_mp3     61159  watching      9              0   
74626630    Door_mp3     61159  watching      9              0   
74566332   Pavle2009     44511  watching      0              0   
74614909   Pavle2009     44511  watching      0              0   
44417439    W00F1234     59457   on_hold      0              0   
44469797    W00F1234     59457   dropped      3              0   

          num_watched_episodes  
74586609                     8  
74616569                     9  
74594917                     5  
74626628                     6  
74594935                     4  
74626629                     5  
74594941                     4  
74626630                     5  
74566332                     5  
74614909                     7  
44417439                     1  
44469797                     1

In [65]:
# usually we should drop all the occurrence of a duplicate and keep the first
# in this case though, it looks like the latest occurence is the most updated one, contaning more info than the first one, so we drop the first one
ratings_df.drop_duplicates(subset=['username', 'anime_id'], keep='last', inplace=True)

We had just 6 duplicates having the same username and anime_id

In [66]:
# we check for Nan values.
#TODO
# if it is necessary check if the num_watched_episodes is greater than number of episodes of anime, we can remove the Nan values and put one or zero. 
ratings_df.isna().sum()

username                      7
anime_id                      0
status                        0
score                         0
is_rewatching           3797321
num_watched_episodes          0
dtype: int64

In [67]:
# drop "username" with Nan values?
#TODO
ratings_df[ratings_df['username'].isna()]

username  anime_id         status  score  is_rewatching  \
100051758      NaN      1482       watching      9              0   
100051759      NaN      1735       watching      9              0   
100051760      NaN       121      completed     10              0   
100051761      NaN       136      completed      8              0   
100051762      NaN       269        on_hold      7              0   
100051763      NaN      1818        on_hold      7              0   
100051764      NaN      1535  plan_to_watch      6              0   

           num_watched_episodes  
100051758                    35  
100051759                    21  
100051760                    51  
100051761                    62  
100051762                    33  
100051763                    11  
100051764                     4

Check this username that there is in the profiles_df

### NINTH DATASET

In [68]:
characters_df

character_mal_id                                                url  \
0               280386.0  https://myanimelist.net/character/280386/Envi_...   
1               280354.0    https://myanimelist.net/character/280354/Eleven   
2               280353.0      https://myanimelist.net/character/280353/Stud   
3               280352.0     https://myanimelist.net/character/280352/Judge   
4               280339.0  https://myanimelist.net/character/280339/Eiji_...   
...                  ...                                                ...   
209958          282276.0  https://myanimelist.net/character/282276/Farra...   
209959          282277.0  https://myanimelist.net/character/282277/Harri...   
209960          282278.0       https://myanimelist.net/character/282278/Rob   
209961          282281.0     https://myanimelist.net/character/282281/Grimm   
209962          282284.0  https://myanimelist.net/character/282284/Joel_...   

                      name       name_kanji  \
0       Envi Mel Champagne  エンヴィ・メル・シャンパーニュ   
1                   Eleven             イレヴン   
2                     Stud             スタッド   
3                    Judge             ジャッジ   
4            Eiji Kurokawa            黒川 英治   
...                    ...              ...   
209958  Farrah Van Dorothy     ファラ・ヴァン・ドロシー   
209959         Harris Mead          ハリス・ミード   
209960                 Rob               ロブ   
209961               Grimm              グリム   
209962       Joel Widdison             ジョエル   

                                                    image  favorites about  
0       https://cdn.myanimelist.net/images/characters/...        0.0   NaN  
1       https://cdn.myanimelist.net/images/characters/...        0.0   NaN  
2       https://cdn.myanimelist.net/images/characters/...        0.0   NaN  
3       https://cdn.myanimelist.net/images/characters/...        0.0   NaN  
4       https://cdn.myanimelist.net/img/sp/icon/apple-...        0.0   NaN  
...                                                   ...        ...   ...  
209958  https://cdn.myanimelist.net/images/characters/...        0.0   NaN  
209959  https://cdn.myanimelist.net/images/characters/...        0.0   NaN  
209960  https://cdn.myanimelist.net/images/characters/...        0.0   NaN  
209961  https://cdn.myanimelist.net/images/characters/...        0.0   NaN  
209962  https://cdn.myanimelist.net/images/characters/...        0.0   NaN  

[209963 rows x 7 columns]

In [69]:
# check types of dataset columns
characters_df.dtypes

character_mal_id    float64
url                  object
name                 object
name_kanji           object
image                object
favorites           float64
about                object
dtype: object

In [70]:
# change "character_mal_id" and "favorites" from float to int
characters_df["character_mal_id"] = characters_df["character_mal_id"].astype("Int64")
characters_df["favorites"] = characters_df["favorites"].astype("Int64")

In [71]:
characters_df.describe()

character_mal_id    favorites
count          209961.0     209961.0
mean      150579.010283    57.616138
std        85243.624423  1197.586488
min                 1.0          0.0
25%             74581.0          0.0
50%            160191.0          0.0
75%            225373.0          2.0
max            282284.0     175632.0

In [72]:
# we have only 2 rows where all columns are Nan, the rows with Nan values in "name_kanji" and "about" we shouldn't drop because they have other values that are important.
characters_df.isna().sum()

character_mal_id        2
url                     2
name                    2
name_kanji          55480
image                   2
favorites               2
about               96976
dtype: int64

In [73]:
# here we want to check if the Nan values are concentrate in only two rows
characters_df[characters_df['character_mal_id'].isna()]

character_mal_id  url name name_kanji image  favorites about
209733              <NA>  NaN  NaN        NaN   NaN       <NA>   NaN
209734              <NA>  NaN  NaN        NaN   NaN       <NA>   NaN

In [74]:
# Apart "name_kanji" and "about" the others Nan values are concentrate in two rows so we drop the two rows with all columns Nan
characters_df.dropna(how='all', inplace=True)


In [75]:
# we want to see all duplicates to understand if we have to drop or not
characters_df.loc[characters_df.duplicated(subset=['character_mal_id', 'url', 'name'], keep=False)]

character_mal_id                                                url  \
853               279200  https://myanimelist.net/character/279200/Shing...   
854               279200  https://myanimelist.net/character/279200/Shing...   
909               279142      https://myanimelist.net/character/279142/Kaho   
910               279142      https://myanimelist.net/character/279142/Kaho   
1166              278865  https://myanimelist.net/character/278865/Yoshi...   
...                  ...                                                ...   
209789            282042     https://myanimelist.net/character/282042/Alina   
209849            282115    https://myanimelist.net/character/282115/Cavitt   
209850            282115    https://myanimelist.net/character/282115/Cavitt   
209858            282131  https://myanimelist.net/character/282131/Lieut...   
209859            282131  https://myanimelist.net/character/282131/Lieut...   

                     name name_kanji  \
853      Shingo Shimazaki      島崎 真悟   
854      Shingo Shimazaki      島崎 真悟   
909                  Kaho        NaN   
910                  Kaho        NaN   
1166    Yoshirou Sonozaki      園崎 義郎   
...                   ...        ...   
209789              Alina        アリナ   
209849             Cavitt      キャビット   
209850             Cavitt      キャビット   
209858  Lieutenant Rencon     レンコン中尉   
209859  Lieutenant Rencon     レンコン中尉   

                                                    image  favorites  \
853     https://cdn.myanimelist.net/images/characters/...          1   
854     https://cdn.myanimelist.net/images/characters/...          1   
909     https://cdn.myanimelist.net/images/characters/...          0   
910     https://cdn.myanimelist.net/images/characters/...          0   
1166    https://cdn.myanimelist.net/images/characters/...          0   
...                                                   ...        ...   
209789  https://cdn.myanimelist.net/images/characters/...          0   
209849  https://cdn.myanimelist.net/images/characters/...          0   
209850  https://cdn.myanimelist.net/images/characters/...          0   
209858  https://cdn.myanimelist.net/images/characters/...          0   
209859  https://cdn.myanimelist.net/images/characters/...          0   

                                                    about  
853                                                   NaN  
854                                                   NaN  
909                                                   NaN  
910                                                   NaN  
1166              The owner of the restaurant Angel Mort.  
...                                                   ...  
209789                                                NaN  
209849  Lieutenant Rencon's younger sister.\n\nNo voic...  
209850  Lieutenant Rencon's younger sister.\n\nNo voic...  
209858  Former lieutenant of Mar Expedition and older ...  
209859  Former lieutenant of Mar Expedition and older ...  

[794 rows x 7 columns]

In [76]:
# we drop the duplicates because they have all same values 
characters_df.drop_duplicates(subset=['character_mal_id', 'url', 'name'], keep='first', inplace=True)

### TENTH DATASET

In [77]:
# role of character anime
character_anime_works_df

anime_mal_id  character_mal_id    character_name        role
0               2928              5781             Atoli        Main
1               2928                33             Haseo        Main
2               2928                32              Ovan        Main
3               2928                34             Shino        Main
4               2928              5785              Aina  Supporting
...              ...               ...               ...         ...
236811         31245            137157    Shibasaki, Ken  Supporting
236812         36305            136064  Hamanaka, Midori        Main
236813         36305            133916      Narumi, Sena        Main
236814         36305            124942   Hayasaka, Akari  Supporting
236815         34895            155887  Shiraha, Setsuna        Main

[236816 rows x 4 columns]

In [78]:
# check types of columns
character_anime_works_df.dtypes

anime_mal_id         int64
character_mal_id     int64
character_name      object
role                object
dtype: object

In [79]:
# check the number of Nan value
character_anime_works_df.isna().sum()

anime_mal_id        0
character_mal_id    0
character_name      0
role                0
dtype: int64

In [80]:
# check the number of duplicates
character_anime_works_df.loc[character_anime_works_df.duplicated(subset=['anime_mal_id', 'character_mal_id'])]

Empty DataFrame
Columns: [anime_mal_id, character_mal_id, character_name, role]
Index: []

There is no need to clean this dataset 

### ELEVENTH DATASET

In [81]:
person_voice_works_df

person_mal_id        role  anime_mal_id  character_mal_id  language
0                   1        Main         55830              2514  Japanese
1                   1  Supporting         60602              2822  Japanese
2                   1  Supporting         59229            140499  Japanese
3                   1  Supporting         60427            275856  Japanese
4                   1  Supporting         62067            190335  Japanese
...               ...         ...           ...               ...       ...
489511          89839  Supporting         40111            274622  Mandarin
489512          89840  Supporting         60544            266412  Mandarin
489513          89841  Supporting         60544            266416  Mandarin
489514          89842  Supporting         36896            279922  Japanese
489515          89952  Supporting         60260            279778  Japanese

[489516 rows x 5 columns]

In [82]:
person_voice_works_df['language'].value_counts()

language
Japanese           203537
English             94902
French              43340
Spanish             39153
Portuguese (BR)     37100
Italian             30712
German              26805
Korean               7981
Mandarin             3529
Hungarian            1378
Hebrew               1079
Name: count, dtype: int64

In [83]:
person_voice_works_df['language'].sum()

: 

: 

In [ ]:
#I want to group by language and see the different languages available in the dataset.
person_voice_works_df.groupby('language').size()

In [ ]:
person_voice_works_df.dtypes

In [ ]:
person_voice_works_df.isna().sum()

In [ ]:
# check if the duplicates are in all columns
person_voice_works_df.loc[person_voice_works_df.duplicated(keep=False)]

In [ ]:
# drop the duplicates because they have all same values
person_voice_works_df.drop_duplicates(keep='first', inplace=True)

### TWELFTH DATASET

In [ ]:
# Should we delete the last five columns?
# TODO
profiles_df

username  gender      birthday       location        joined  \
0         ishikawas     NaN           NaN    South Korea           NaN   
1              CKK2     NaN           NaN  United States   Dec 1, 2018   
2       --------788  Female           NaN         Mexico   Oct 4, 2022   
3        potatoaris     NaN           NaN          Spain   Oct 2, 2018   
4         Rinrintan     NaN           NaN          Japan  May 12, 2019   
...             ...     ...           ...            ...           ...   
337150  ariyanroy04     NaN           NaN  United States  Mar 30, 2021   
337151   ariyanvk18    Male           NaN         Turkey  Jan 24, 2022   
337152     ariyoskz     NaN           NaN        Germany   Jul 7, 2022   
337153    arizima23  Female  Jun 23, 1998          Spain  Mar 15, 2019   
337154      arizkim     NaN           NaN   South Africa   Dec 7, 2020   

       watching completed on_hold dropped plan_to_watch  
0           NaN       NaN     NaN     NaN           NaN  
1             3       182      15       0           405  
2             1        64       0       0             1  
3             5         1       0       0             4  
4            20       311      40      16            34  
...         ...       ...     ...     ...           ...  
337150       35       130       0       0           489  
337151        7       101       1      12            21  
337152        1       155       2      11            15  
337153       15        64       0       4             4  
337154        7        38       1       0            33  

[337155 rows x 10 columns]

In [ ]:
# check if the types are right for each field
profiles_df.dtypes

In [ ]:
# trying things with the date
# BEFORE running the .to_datetime command, ishiyama_yumi has Jul 24 bday
profiles_df.loc[44]

username         ishiyama_yumi
gender                  Female
birthday                Jul 24
location              Thailand
joined            Dec 16, 2020
watching                    10
completed                  308
on_hold                     16
dropped                     56
plan_to_watch              143
Name: 44, dtype: object

In [ ]:
# change types of columns "birthday" and "joined" from object to date and the others columns that they should be int
profiles_df["birthday"] = pd.to_datetime(profiles_df["birthday"], errors='coerce')
profiles_df["joined"] = pd.to_datetime(profiles_df["joined"], errors='coerce')

In [ ]:
# AFTER running the .to_datetime command, ishiyama_yumi has NaT birthday
profiles_df.loc[44]

username               ishiyama_yumi
gender                        Female
birthday                         NaT
location                    Thailand
joined           2020-12-16 00:00:00
watching                          10
completed                        308
on_hold                           16
dropped                           56
plan_to_watch                    143
Name: 44, dtype: object

In [ ]:
profiles_df["birthday"].min(), profiles_df["birthday"].max()


(Timestamp('1800-01-01 00:00:00'), Timestamp('2100-12-31 00:00:00'))

In [ ]:
weird_birthdays = profiles_df[
    (profiles_df["birthday"] < "1900-01-01") |
    (profiles_df["birthday"] > "2025-12-31")
]

weird_birthdays


username  gender      birthday   location        joined watching  \
15           RinsAl    Male   Jul 7, 1989      Japan  Oct 19, 2009       16   
19         ishimori  Female  Apr 17, 2000     France   Dec 7, 2018        1   
23       Karinnn___     NaN  Feb 28, 2001  Australia   Aug 7, 2018        2   
25            Rins_    Male  Jul 15, 1988      Japan   Jun 6, 2013        5   
29      FollowValen    Male   Jul 8, 1996      Spain   Oct 7, 2018        9   
...             ...     ...           ...        ...           ...      ...   
337137    arisvgawa  Female   Mar 3, 2001      Japan   Oct 2, 2024        1   
337138      aritaka     NaN         Aug 4    Germany  Apr 24, 2012       18   
337143     aritsune     NaN  Oct 27, 1998      Japan  Jul 20, 2012        4   
337147       ariu82    Male  Sep 17, 2003      China  Dec 25, 2020        1   
337153    arizima23  Female  Jun 23, 1998      Spain  Mar 15, 2019       15   

       completed on_hold dropped plan_to_watch  
15           484       5      19            70  
19            91      16      14            30  
23           380      11      36            62  
25           110       0       1            46  
29            48       9       5            33  
...          ...     ...     ...           ...  
337137        16       0       0             0  
337138       466       3      48            33  
337143        20       3       1            20  
337147       228      15      12            29  
337153        64       0       4             4  

[115390 rows x 10 columns]

In [ ]:
#todo
# Just noticed some birthdays have no year but just day and month, how could we manage those?
# I guess with errors='coerce', when we change type from object to date, the year is set to a default one (e.g. 1800 or 1900)

In [ ]:
profiles_df.loc[
    profiles_df["birthday"] > profiles_df["joined"],
    ["birthday", "joined"]
]


birthday     joined
2606   2021-07-13 2020-09-14
2683   2019-07-11 2017-10-13
2817   2077-07-18 2022-01-15
4506   2013-09-03 2013-01-26
4616   2017-01-12 2015-01-30
...           ...        ...
333376 2022-09-05 2019-11-22
334143 2019-10-11 2012-04-21
334335 2020-04-04 2018-12-31
335340 2023-09-05 2021-09-12
335927 2022-06-15 2016-08-26

[284 rows x 2 columns]

In [ ]:
profiles_df["joined"].min(), profiles_df["joined"].max()

(Timestamp('2004-11-05 00:00:00'), Timestamp('2025-11-02 00:00:00'))

In [ ]:
profiles_df["birthday"].dt.year.value_counts().sort_index().head(30)

In [ ]:
profiles_df[profiles_df["birthday"].dt.year == 1930]

In [ ]:
# decided to remove the birthdays of people older than 100 when they joined the website
mask_too_old = (profiles_df["joined"] - profiles_df["birthday"]).dt.days / 365.25 > 100
profiles_df.loc[mask_too_old, "birthday"] = pd.NaT


In [ ]:
mask_too_young = (profiles_df["joined"] - profiles_df["birthday"]).dt.days / 365.25 < 3
mask_too_young


0         False
1         False
2         False
3         False
4         False
          ...  
337150    False
337151    False
337152    False
337153    False
337154    False
Length: 337155, dtype: bool

In [ ]:
profiles_df.isna().sum()
#TODO one guy has null username
# We'll put a progressive unkown_user[] label on it, so that for future data with missing username we could go on with it

username              1
gender           170876
birthday         251156
location              0
joined             1676
watching           1678
completed          1678
on_hold            1678
dropped            1678
plan_to_watch      1678
dtype: int64

In [ ]:
# check if there is any duplicate on "username"
profiles_df.loc[profiles_df.duplicated(subset=['username'], keep='first')]
# none found

Empty DataFrame
Columns: [username, gender, birthday, location, joined, watching, completed, on_hold, dropped, plan_to_watch]
Index: []

### THIRTEENTH DATASET

In [ ]:
recommendations_df

In [ ]:
recommendations_df.dtypes

In [ ]:
recommendations_df.isna().sum()

In [ ]:
recommendations_df.loc[profiles_df.duplicated(keep='first')]

# TO KEEP

In [ ]:
# 1st
# dropping duplicates
character_nicknames_df.drop_duplicates(keep='first', inplace=True)
# dropping nan values
character_nicknames_df.dropna(inplace=True)

# 2nd
# no need to clean from duplicates nor missing values
# change column types from object to datetime
details_df[["start_date", "end_date"]] = details_df[["start_date", "end_date"]].apply(
    pd.to_datetime, errors="coerce"
)
# change column types from object to Int64
details_df[["scored_by", "rank", "episodes", "year"]] = (
    details_df[["scored_by", "rank", "episodes", "year"]].astype("Int64")
)

# 3rd
# no need to clean


# 4th
#dropping nan values because they don't give any useful information
person_alternate_names_df.dropna(inplace=True)
# dropping duplicates
person_alternate_names_df.drop_duplicates(keep='first', inplace=True)


# 5th
# Dropping duplicates choosing to keep the first occurrence. They only differ for the "relevant_location" field which has no interest for us
person_details_df.drop_duplicates(subset=['person_mal_id', 'url', 'name'], keep='first', inplace=True)
# change column types from object to datetime
person_details_df["birthday"] = pd.to_datetime(person_details_df["birthday"], errors='coerce')

# There are two weird rows with nan values in most columns but we don't drop them because they may be linked to other datasets
# Could we join the two tables person_details_df and person_alternate_names_df having the keys that match.
# Putting the alternate names in a new column called alt_name and having a list of those inside
# TODO


# 6th
# no need to clean


# 7th
# change column types to save memory
stats_df.filter(regex="_votes$").astype("Int64")
stats_df[stats_df.filter(regex="_votes$").columns] = (
    stats_df.filter(regex="_votes$").astype("Int64")
)


# 8th
# change column types to save memory
ratings_df["is_rewatching"] = ratings_df["is_rewatching"].astype("Int8")
ratings_df["anime_id"] = ratings_df["anime_id"].astype("Int32")
# drop duplicates keeping the last entry (most recent)
ratings_df.drop_duplicates(subset=['username', 'anime_id'], keep='last', inplace=True)
# there's a username with NaN value, we'll keep it for now as it matches with the profiles_df
# we only have 1 Nan value in "username" in the profiles_df so we can keep it for now
# I could manage it this way: generate a deterministic placeholder, e.g.:
# the maximum existing user_id + 1
# or a specific labeled ID like "unknown_user"
# THEN we can use this placeholder consistently across all datasets to maintain referential integrity.
# profiles_df["user_id"] = profiles_df["user_id"].fillna(new_id)
# ratings_df["user_id"] = ratings_df["user_id"].fillna(new_id)
# TODO


# 9th
# change "character_mal_id" and "favorites" from float to int
characters_df[["character_mal_id", "favorites"]] = (
    characters_df[["character_mal_id", "favorites"]].astype("Int64")
)

# we drop the two rows with all columns Nan
characters_df.dropna(how='all', inplace=True)
# we drop the duplicates because they have all same values 
characters_df.drop_duplicates(subset=['character_mal_id', 'url', 'name'], keep='first', inplace=True)


# 10th
# no need to clean


# 11th
person_voice_works_df.drop_duplicates(keep='first', inplace=True)


# 12th
# change column types from object to datetime
profiles_df[["birthday", "joined"]] = profiles_df[["birthday", "joined"]].apply(
    pd.to_datetime, errors="coerce"
)
# setting birthdays outside a reasonable range to NaT
profiles_df.loc[
    profiles_df["birthday"] > profiles_df["joined"],
    "birthday"
] = pd.NaT
# decided to remove the birthdays of people older than 100 at join
mask_too_old = (profiles_df["joined"] - profiles_df["birthday"]).dt.days / 365.25 > 100
profiles_df.loc[mask_too_old, "birthday"] = pd.NaT
#I mean MongoDB, a female born in 1930-07-09, from Thailand and joined in 2017-07-09 makes perfect sense, doesn't it?

# 13th
# no need to clean


### QUESTIONS:
1. Can we just leave the commands to clean the dataset instead of leaving all the exploration commands such as .describe, dtypes and so on
One file with data cleaning, showing the reasoning.
ANOTHER file with Data Visualization.

2. Ask for the maximum age in the 5th dataset
Justify why you have changed the dates (e.g. change 2070 in 1970) or nan. Possiamo fare il cazzo che vogliamo

3. Ask for the only NaN username in the "ratings" dataset: for the hci part?
Think in terms of the future, take the latest id and sum 1, or progressive uknownuserssssssss (e.g. uknown1, unknown2 in caso poi riceviamo altri dati nel futuro)

For TWEB we could just put a random (?) value in order not to lose the line and have non nan values in the database

SE NON USIAMO COLONNE IN ASSOLUTO, per alcuna query o nel sito tweb
Siccome non faremo mai query su questa cosa qui POSSIAMO TOGLIERE le colonne (e.g. kanji name lo possiamo togliere)